server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -546,7 +546,7 @@ public ParsedDocument index(IndexItem item) throws IOException {
                 ValueIndexer<Object> valueIndexer = (ValueIndexer<Object>) valueIndexers.get(i);
                 xContentBuilder.field(reference.column().leafName());
                 valueIndexer.indexValue(
-                    value,
+                    reference.valueType().sanitizeValue(value),
```

Comment:
Shouldn't we rather add custom implementations of `DataType.valueForInsert` for Long and Doubles (the only numbers our parser knows)?
Adding an additional sanitization feels wrong to me.

Reply:
Afaik this is all about inserting values (and replicate the inserts), so I do not really understand why `valueForInsert` shouldn't fit.

But I really do not understand the root issue, if there is some sanitization missing, why doesn't it fail on the primary as well? Also why does the replication work even if the exception is thrown?
Lets please find answers to this before we make any change only based on assumptions.

Reply:
> why doesn't it fail on the primary as well?

On primary, since it is a first insert to the cluster, `DynamicIndexer` is used which calls `guessType` then `upcast` that converts `Integer` to `Long`, while insert to replica is a second insert to the cluster, it is no longer `DynamicIndexer` but a `LongIndexer` which causes `ClassCastException`.

> Also why does the replication work even if the exception is thrown?
I am looking into this.

Reply:
> Also why does the replication work even if the exception is thrown?

It looks to me that `PeerRecoveryTargetService` is handling the failed replica. It goes through methods like `IndexShard.applyTranslogOperation` -> `DocumentParser.parseDocument` which is why the replication completes eventually.

Reply:
@jeeminso Make sense, thanks for diving into this.
In general I think we should discuss if the `upcast` for `integer` types is really what we want, given that our parser will detect and return integer types (see https://github.com/crate/crate/pull/10066). I'd tend to not upcast integer values. @mfussenegger WDYT?

Reply:
> In general I think we should discuss if the upcast for integer types is really what we want, given that our parser will detect and return integer types

I wouldn't change the behavior for that. It's possible that only the _first_ value we receive fits into an integer, and a latter wouldn't.

Reply:
How should we proceed from here? @seut @mfussenegger

Reply:
Both options (using sanitizeValue in addition, or adding `valueForInsert` implementations for Long/Integer/Short types seem okay to me for the fix.

We can follow up and narrow their expected use if appropriate.

Maybe it is even possible to merge the two, but that's out of scope for a bugfix and not that urgent. Maybe something for the feature freeze period.

Reply:
> We can follow up and narrow their expected use if appropriate.

>Maybe it is even possible to merge the two, but that's out of scope for a bugfix and not that urgent. Maybe something for the feature freeze period.

Opened https://github.com/crate/crate/issues/14178 to track this, I will have a final look and merge.